mobile_verifier/src/boosting_oracles/mod.rs

```diff
@@ -1,184 +1,600 @@
-            90 => Landtype::Wet,
-            95 => Landtype::Mangrove,
-            100 => Landtype::Moss,
-            other => anyhow::bail!("unexpected landtype disktree value: {other:?}"),
-        };
-        Ok(val)
+impl<Foot, Land, Urban> HexBoostData<Foot, Land, Urban> {
+    pub fn new(footfall: Foot, landtype: Land, urbanization: Urban) -> Self {
+        Self {
+            urbanization: Arc::new(Mutex::new(urbanization)),
+            footfall: Arc::new(Mutex::new(footfall)),
+            landtype: Arc::new(Mutex::new(landtype)),
+        }
+    }
+}
+
+impl<Foot, Land, Urban> HexBoostData<Foot, Land, Urban>
+where
+    Foot: DataSet,
+    Land: DataSet,
+    Urban: DataSet,
+{
+    pub async fn is_ready(&self) -> bool {
+        self.urbanization.lock().await.is_ready()
+            && self.footfall.lock().await.is_ready()
+            && self.landtype.lock().await.is_ready()
+    }
+}
+
+impl<Foot, Land, Urban> HexBoostData<Foot, Land, Urban>
+where
+    Foot: HexAssignment,
+    Land: HexAssignment,
+    Urban: HexAssignment,
+{
+    pub async fn assignments(&self, cell: hextree::Cell) -> anyhow::Result<HexAssignments> {
+        let footfall = self.footfall.lock().await;
+        let landtype = self.landtype.lock().await;
+        let urbanization = self.urbanization.lock().await;
+        HexAssignments::from_data_sets(cell, &*footfall, &*landtype, &*urbanization)
+    }
+}
+
+trait DiskTreeLike: Send + Sync {
+    fn get(&self, cell: hextree::Cell) -> hextree::Result<Option<(hextree::Cell, &[u8])>>;
+}
+
+impl DiskTreeLike for DiskTreeMap {
+    fn get(&self, cell: hextree::Cell) -> hextree::Result<Option<(hextree::Cell, &[u8])>> {
+        self.get(cell)
     }
 }
 
+impl DiskTreeLike for std::collections::HashSet<hextree::Cell> {
+    fn get(&self, cell: hextree::Cell) -> hextree::Result<Option<(hextree::Cell, &[u8])>> {
+        Ok(self.contains(&cell).then_some((cell, &[])))
+    }
+}
+
+pub struct MockDiskTree;
+
+impl DiskTreeLike for MockDiskTree {
+    fn get(&self, cell: hextree::Cell) -> hextree::Result<Option<(hextree::Cell, &[u8])>> {
+        Ok(Some((cell, &[])))
+    }
+}
+
+#[derive(FromRow)]
+pub struct UnassignedHex {
+    uuid: uuid::Uuid,
+    #[sqlx(try_from = "i64")]
+    hex: u64,
+    signal_level: SignalLevel,
+    signal_power: i32,
+}
+
+impl UnassignedHex {
+    pub fn fetch_all(pool: &PgPool) -> impl Stream<Item = sqlx::Result<Self>> + '_ {
+        sqlx::query_as("SELECT uuid, hex, signal_level, signal_power FROM hexes").fetch(pool)
+    }
+
+    pub fn fetch_unassigned(pool: &PgPool) -> impl Stream<Item = sqlx::Result<Self>> + '_ {
+        sqlx::query_as(
+            "SELECT
+                uuid, hex, signal_level, signal_power
+            FROM
+                hexes
+            WHERE
+                urbanized IS NULL
+                OR footfall IS NULL
+                OR landtype IS NULL",
+        )
+        .fetch(pool)
+    }
+}
+
+pub async fn set_oracle_boosting_assignments<'a>(
```

Comment:
this function would be more palatable with some elements broken out

Reply:
sure, but this function has been in the verifier since urbanization dropped